engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/modelconfig/ModelConfigLoader.scala

```diff
@@ -0,0 +1,63 @@
+package pl.touk.nussknacker.engine.modelconfig
+
+import com.typesafe.config.{Config, ConfigFactory}
+import pl.touk.nussknacker.engine.modelconfig.ModelConfigLoader.defaultModelConfigResource
+
+object ModelConfigLoader {
+
+  private val defaultModelConfigResource = "defaultModelConfig.conf"
+
+}
+
+/**
+  * Loads model configs.
+  *
+  * Configs are resolved once when loading model data. Implement this trait when you want to add dynamic configuration
+  * that is fetched from external source, for example OpenAPI specification and you would like this configuration to
+  * be immutable for deployed processes.
+  *
+  * This trait is optional to implement.
+  */
+abstract class ModelConfigLoader extends Serializable {
+
+  /**
+    * Resolves config part to pass in execution during e.g. process deployment on Flink cluster. When running
+    * a process (e.g. on Flink) resolved config part is used to construct full config.
+    *
+    * Method used for performance reasons to reduce serialized configuration size inside deployed processes. By default
+    * config from main nussknacker file at path: processTypes.{type_name}.modelConfig is used.
+    *
+    * @param configDuringModelAnalysis default config resolved by [[resolveConfigDuringExecution]] for configuration
+    *                                  from processTypes.{type_name}.modelConfig
+    * @return transformation of config part (processTypes.{type_name}.modelConfig) that is later passed to a running
+    *         process (see e.g. FlinkProcessCompiler, ModelConfigToLoad).
+    */
+  def resolveConfigToPassInExecution(configDuringModelAnalysis: Config, classLoader: ClassLoader): ConfigToPassInExecution
+
+  /**
+    * Resolves full config used inside [[pl.touk.nussknacker.engine.api.process.ProcessConfigCreator]]. Invoked both
+    * in NK and running process.
+    *
+    * Default implementation loads model config the from following locations:
+    * <ol>
+    * <li> param configPassedInExecution resolved by [[resolveConfigToPassInExecution]], by default part of main config
+    * file of nussknacker (in path processTypes.{type_name}.modelConfig)</li>
+    * <li>defaultModelConfig.conf from model jar</li>
+    * <li>reference.conf from model jar</li
+    * </ol
+    *
+    * Default implementation should rather not be changed.
+    */
+  def resolveConfigDuringExecution(configPassedInExecution: Config, classLoader: ClassLoader): Config = {
```

Comment:
Can you extract new protected method e.g. `resolveConfigUsingDefaults` which will be invoked by this method? IMO it will make the usage of this method more easy to understand in concrete implementations.